base/strings/annotated.jl

```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
 Get all annotations of `chr`.
 """
 annotations(c::AnnotatedChar) = c.annotations
+
+## AnnotatedIOBuffer
+
+struct AnnotatedIOBuffer <: AbstractPipe
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+end
+
+AnnotatedIOBuffer(io::IOBuffer) = AnnotatedIOBuffer(io, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+AnnotatedIOBuffer() = AnnotatedIOBuffer(IOBuffer())
+
+function show(io::IO, aio::AnnotatedIOBuffer)
+    show(io, AnnotatedIOBuffer)
+    print(io, '(', aio.io.size, " byte", ifelse(aio.io.size == 1, "", "s"), ", ",
+          length(aio.annotations), " annotation", ifelse(length(aio.annotations) == 1, "", "s"), ")")
+end
+
+pipe_reader(io::AnnotatedIOBuffer) = io.io
+pipe_writer(io::AnnotatedIOBuffer) = io.io
+
+# Useful `IOBuffer` methods that we don't get from `AbstractPipe`
```

Comment:
Not to hold up this PR, but do you think we should add these to AbstractPipe also (all on the read side), except `copy` which doesn't have a generic implementation, or is it too much ambiguous whether this intends to attempt to manipulate the reader or writer half

Reply:
This is a slightly tricky one I think. I've found the lack of `position` and `seek` annoying with `IOContext`-wrapped `IOBuffer`s myself, however I think the problem is that sometimes IO has a read-head and a write-head, and sometimes it doesn't.

I think the cleaner way for this to be implemented is to have read-head and write-head variants of `position`/`seek`/etc. and if needed have them refer to the same head for IO types where there can only be one head. However, instead the default assumption is that they are the same, and so the model falls short as soon as there is distinct input and output IO.